OS Managed Power Management framework

ramakrishnapallala opened the pull request:
arch: arm: nordic_nrf: Add an API to check for valid PM state
subsys: power: Add support for OS managed Power Management framework
samples: subsys: Add sample app to demo OS managed PM

Fixes #8255

Commit:
arch: arm: nordic_nrf: Add an API to check for valid PM state

Add an API _sys_soc_is_valid_power_state to check if a PM state
is valid or not.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

dbkinder commented on samples/boards/nrf52/power_mgr_os/README.rst:
```diff
@@ -0,0 +1,92 @@
+.. _nrf52-power-mgr-sample:
```
This looks like it's a copy of an existing sample documentation with no obvious changes (including the title):  ``/samples/boards/nrf52/power_mgr/README.rst``.    The doc build complained that the label at the top of the doc here was the same as another document.

ramakrishnapallala replied:
yes David, will fix it.

dbkinder commented on samples/boards/nrf52/power_mgr_os/README.rst:
```diff
@@ -0,0 +1,92 @@
+******************************
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/boards/nrf52/power_mgr
```
... and the sample app you're running is in another folder, not the one with this documentation in it.

ramakrishnapallala replied:
Ok will fix it.

ramakrishnapallala commented on arch/arm/soc/nordic_nrf/nrf52/power.c:
```diff
@@ -100,3 +100,23 @@ void _sys_soc_power_state_post_ops(enum power_states state)
+#if defined(CONFIG_SYS_POWER_DEEP_SLEEP)
+	case SYS_POWER_STATE_DEEP_SLEEP:
+#endif
+		lps_flag = true;
```
Yes will do.

ramakrishnapallala commented on arch/arm/soc/nordic_nrf/nrf52/power.c:
```diff
@@ -100,3 +100,23 @@ void _sys_soc_power_state_post_ops(enum power_states state)
+		lps_flag = true;
+		break;
+	default:
+		lps_flag = false;
```
yes, will do.

ramakrishnapallala commented on samples/boards/nrf52/power_mgr_os/README.rst:
```diff
@@ -0,0 +1,92 @@
+   transparent to devices. SOC and devices do not lose context in this Mode.
+   SOC supports the following two Low Power states :
+
+   A. CONSTANT LATENCY LOW POWER MODE : This Low Power State triggers CONSTLAT
```
@pizi-nordic this is sample application documentation which i reused from existing sample. I will double check and fix the gaps.

Also on the PM states, from the documentation i saw that the nrf5x has support for two active power saving states (constant latency mode and low power mode).

The idea is to support pluggable PM policies based on the target application, in this base PR, I have added a simple policy to demonstrate the overall flows.

pizi-nordic replied:
Are you referring the "Sub power modes" in http://infocenter.nordicsemi.com/index.jsp?topic=%2Fcom.nordic.infocenter.nrf52832.ps.v1.1%2Fpower.html ?

If so, you see that mentioned power modes only affect PPI interconnect delay and CPU wakeup time. Changing these settings based on predicted sleep time, which is done in the example, is not a thing which we can recommend as these settings are in fact driven by the real-time requirements of application (we might break operation of PPI-based solutions relying on constant event latency if we switch to low power mode in runtime).

Despite of the fact, that this feature is a bit hard to show in nRF5x family, I think, that OS-level power management is a important feature which have to be implemented in Zephyr.

ramakrishnapallala commented on subsys/power/Kconfig:
```diff
@@ -0,0 +1,101 @@
+config PM_CONTROL_OS_LPS
+	bool "Platfrom supports LPS"
+	help
+	  This is the maximum number of bytes that the
```
copy paste error, will fix it.

ramakrishnapallala commented on subsys/power/policy.c:
```diff
@@ -0,0 +1,101 @@
+#include <soc.h>
+#include <soc_power.h>
+
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_PM_LEVEL
```
Ok will fix it.

ramakrishnapallala commented on subsys/power/device.c:
```diff
@@ -0,0 +1,104 @@
+ * and clock domain dependencies.
+ */
+#ifdef CONFIG_SOC_SERIES_NRF52X
+#define CONFIG_MAX_PM_DEVICES	15
```
Ok will consider using the device list itself.

ramakrishnapallala commented on subsys/power/device.c:
```diff
@@ -0,0 +1,104 @@
+		if (!device_retval[i]) {
+			int idx = device_ordered_list[i];
+
+			device_set_power_state(&pm_device_list[idx],
```
I think this kind of complexity can be handled in the driver itself, for example, on resume driver can submit worker to do the actual wake up and handle the device status accordingly.

pizi-nordic replied:
Ok. This will also make the sleep state longer saving more time. What about applications which need small latency?

ramakrishnapallala commented on subsys/power/device.c:
```diff
@@ -0,0 +1,104 @@
+			}
+		}
+
+		if (is_core_dev) { 
```
As you pointed rightly, we need to use device tree here to handle the device dependencies which will come as incremental changes in the upcoming patches.

Commit:
subsys: power: Add OS managed Power Management framework

Add support for OS managed Power Management framework for Zephyr
under 'subsys/power'. This framework takes care of implementing
the _sys_soc_suspend/_sys_soc_resume API's, a PM policy based on
SoC Low Power residencies and also provides necessary API's to
do devices suspend and resume.

Also add necessary changes to support the existing Application
managed Power Management framework.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
samples: subsys: Add sample app to demo OS managed PM

Add a sample application to demonstrate OS managed
Power Management framework.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

ramakrishnapallala commented:
@pizi-nordic  @dbkinder  I have addressed your comments and updated the PR, please have a look.

ramakrishnapallala commented:
@dbkinder I am getting following warning:

/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/doc/_build/rst/doc/samples/subsys/power/power_mgr/README.rst: WARNING: document isn't included in any toctree

How to fix this?

nashif commented:
> How to fix this?

```
diff --git a/samples/subsys/power/power.rst b/samples/subsys/power/power.rst
new file mode 100644
index 0000000000..e1dd2cc6ee
--- /dev/null
+++ b/samples/subsys/power/power.rst
@@ -0,0 +1,10 @@
+.. _power_management-samples:
+
+Power Management Samples
+########################
+
+.. toctree::
+   :maxdepth: 1
+   :glob:
+
+   **/*
```

rgundi commented on subsys/power/pm_policy.h:
```diff
@@ -0,0 +1,57 @@
+extern  void sys_pm_create_device_list(void);
+
+/**
+ * @brief Function to supend the devices in PM device list
```
nitpick - spelling mistake in suspend

rgundi commented on subsys/power/policy.c:
```diff
@@ -0,0 +1,98 @@
+#else
+#define SECS_TO_TICKS		100
+#endif
+
```
Isn't the ticks configurable? So, shouldn't you use a macro function to address it here?

pizi-nordic commented on subsys/power/device.c:
```diff
@@ -0,0 +1,104 @@
+		 * device in the device list is busy as shown below :
+		 * if(device_busy_check(&device_list[idx])) {do something}
+		 */
+		device_retval[i] = device_set_power_state(&pm_device_list[idx],
```
I think, that until blocking still is there, this feature will be not really usable in real-word applications.

pizi-nordic commented:
>SYS_LOG: I looked at the ./doc/subsystems/logging/logger.rst but an example in kernel or driver code will be really helpful.

Please look at https://github.com/zephyrproject-rtos/zephyr/pull/8816. I think that this is the most comprehensive example.  There are also few PRs which are smaller (like https://github.com/zephyrproject-rtos/zephyr/pull/9361), but do not show all features.

nashif commented:
@pizi-nordic are you ok with the changes otherwise? I think the PR was changed to use new logging subsystem already and we can further optimize the usage in the stabilisation period after this gets merged as part of 1.13

pizi-nordic commented:
@pdunaj: FYI

pizi-nordic commented:
@nashif: The biggest issue what I see is bonding of CPU/PPI latency mode to the predicted sleep time. IMHO we should completely remove code which alters this setting from the `nrf52/power.c`.

I assume that rest of the pointed things will be solved later in subsequent PRs, so I have no problem with them.

pizi-nordic commented on subsys/power/policy.c:
```diff
@@ -0,0 +1,98 @@
+#ifdef CONFIG_TICKLESS_KERNEL
+#define SECS_TO_TICKS		CONFIG_TICKLESS_KERNEL_TIME_UNIT_IN_MICRO_SECS
+#else
+#define SECS_TO_TICKS		CONFIG_SYS_CLOCK_TICKS_PER_SEC
```
Such tick <-> time conversion is imprecise on some SoCs (expecially nRF5x). If you need a more precise timing, please use `__ticks_to_ms()` and `_ms_to_ticks()`.

ramakrishnapallala commented:
@pizi-nordic 
"The biggest issue what I see is bonding of CPU/PPI latency mode to the predicted sleep time. IMHO we should completely remove code which alters this setting from the nrf52/power.c."

_sys_soc_set_power_state from nrf52/power.c is only offering a mechanism to enter different power states(SYS_POWER_STATE_CPU_LPS/SYS_POWER_STATE_CPU_LPS_1) and does not have any info on power state residency or latency requirement. It is the PM policy which is making decisions on which PM state to enter through  _sys_soc_suspend which is being defined by the application in the current code. The proposed changes are built on top of them.

So the application designer or developer has to define a PM policy which suits his requirements, like if the system/CPU is going to idle for X secs(residency), then the system can enter into LPS. And these are configurable and can be changed based on application needs.

pdunaj commented on subsys/power/device.c:
```diff
@@ -0,0 +1,115 @@
+	"clk_k32src",
+	"clk_m16src",
+	"sys_clock",
+	"UART_0",
```
UART may be disabled
